server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,230 @@
+         * Otherwise it may attempt to look up values that do not share the correct prefix, which
+         * can result in undefined behavior or an error.
+         */
+        @Override
+        public BytesRef lookupOrd(long ord) throws IOException {
+            BytesRef keyedValue = delegate.lookupOrd(ord);
+            int prefixLength = key.length + 1;
+            int valueLength = keyedValue.length - prefixLength;
+            return new BytesRef(keyedValue.bytes, prefixLength, valueLength);
+        }
+
+        @Override
+        public long nextOrd() throws IOException {
+            if (minOrd < 0) {
+                return NO_MORE_ORDS;
+            }
+
+            for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                if (minOrd <= ord) {
+                    if (ord <= maxOrd) {
+                        return ord;
+                    } else {
+                        return NO_MORE_ORDS;
+                    }
+                }
+            }
+            return NO_MORE_ORDS;
+        }
+
+        @Override
+        public boolean advanceExact(int target) throws IOException {
+            if (minOrd < 0) {
+                return false;
+            }
+
+            if (delegate.advanceExact(target)) {
+                for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                     if (minOrd <= ord && ord <= maxOrd) {
+                         boolean advanced = delegate.advanceExact(target);
+                         assert advanced;
+                         return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /**
+         * Performs a binary search to find the first term with 'key' as a prefix.
+         */
+        static long findMinOrd(BytesRef key, SortedSetDocValues delegate) throws IOException {
+            long low = 0;
+            long high = delegate.getValueCount() - 1;
+
+            long result = -1;
+            while (low <= high) {
+                long mid = (low + high) / 2;
+                final BytesRef term = delegate.lookupOrd(mid);
+                int cmp = compare(key, term);
+
+                if (cmp == 0) {
+                    result = mid;
+                    high = mid - 1;
+                } else if (cmp < 0) {
+                    high = mid - 1;
+                } else {
+                    low = mid + 1;
+                }
+            }
+            return result;
+        }
+
+        /**
+         * Performs a binary search to find the last term with 'key' as a prefix.
+         */
+        static long findMaxOrd(BytesRef key, SortedSetDocValues delegate) throws IOException {
+            long low = 0;
+            long high = delegate.getValueCount() - 1;
+
+            long result = -1;
+            while (low <= high) {
+                long mid = (low + high) / 2;
```

Comment:
Even though it's impossible here because we are dealing with longs, a good practice when implementing binary search with signed indexes is to use unsigned shifts or unsigned division to avoid issues in case of overflow, ie. `(low + high) >>> 1`, or `Long.divideUnsigned(low + high, 2)`.

Reply:
👍